### DISCUSSION TITLE
Extended MAC to IPv6 addrr

### DISCUSSION THREAD

**Discussion by nandra** [openthread/openthread#10729] [General]
Hello, I've go through ot cli interface and I can setup 3 nodes without issue. For reaching node I can run `ipaddr` command to see address and then use ping or other to communicate with node. Is there simple way to do such thing programatically? I know there exists SRP server/client but maybe exist something else? Can I somehow reuse `router table` command output? Seems Extended MAC is somehow similar to node IPaddr? Thanks for any pointers.

**Comment by abtink** [comment]
@nandra, here's a suggestion for you to consider: You can construct an RLOC IPv6 address for a Thread device if you know its RLOC16 (short MAC address). This RLOC IPv6 address can then be used to communicate with the device.  This can help https://openthread.io/guides/thread-primer/ipv6-addressing

**Reply by nandra** [reply]
@abtink I've investigated a bit and I'm still unsure howto do that. I have network with 3 nodes and `router table` shows:
```
> router table

| ID | RLOC16 | Next Hop | Path Cost | LQ In | LQ Out | Age | Extended MAC     | Link |
+----+--------+----------+-----------+-------+--------+-----+------------------+------+
|  2 | 0x0800 |        4 |         1 |     3 |      3 |   7 | eada212c3d2ad6ad |    1 |
|  4 | 0x1000 |        2 |         1 |     3 |      3 |   9 | 9a3bcc248da5696f |    1 |
| 38 | 0x9800 |       63 |         0 |     0 |      0 |   0 | 8631eeb75ab7b176 |    0 |
```
where link = 0 is local device and his address used for communicating is:
```
> ipaddr

fd49:4239:dd86:28:0:ff:fe00:fc00
fd49:4239:dd86:28:0:ff:fe00:9800
fd49:4239:dd86:28:30a1:e19e:2c9:d745
fe80:0:0:0:8431:eeb7:5ab7:b176
```
which more less fits: 8631eeb75ab7b176 and 8431:eeb7:5ab7:b176 (just at the beginning is 86 instead of 84). So Extended MAC can be used to construct node Link-Local Address (LLA)? Is my assumption right please? Thanks.

**Reply by abtink** [reply]
`fd49:4239:dd86:28:0:ff:fe00:9800` is its RLOC address. Note the pattern, `<mesh-local prefix>:0:ff:fe00:<rlco16>`.

You can construct RLOC addresses of other devices if you know their RLOC16, e.g. for the node with RLOC16 of `0x1000` its RLOC Ipv6 address will be `fd49:4239:dd86:28:0:ff:fe00:1000`

**Reply by nandra** [reply]
How to get mesh-local-prefix please? Also according this: https://openthread.io/guides/thread-primer/ipv6-addressing RLOC address shouldn't be used by applications  and I would like to use IP address for communication between nodes? Thanks.

**Reply by jwhui** [reply]
The Mesh Local Prefix is configured as part of the Active Operational Dataset.

- CLI for managing Operational Datasets: https://openthread.io/reference/cli/concepts/dataset
- CLI for directly accessing Mesh Local Prefix: https://openthread.io/reference/cli/commands#prefix_meshlocal

As for whether the RLOC can be directly used, it really depends on what your intent is. The RLOC for a given device is derived based on the topology, so it can change as the Thread network topology changes.

If you're looking for a more general way to discover services and resolve IPv6 addresses, then DNS-SD is the suggested method, which is consistent with how many existing IP-based devices operate today.

**Reply by nandra** [reply]
@jwhui ok I read the comments and now I understand better. But would be possible somehow to get on one node ML-EID address from other node (without using srp etc). basically I want to communicate with nodes so need to somehow know their ip address. When node is added to network isn't information about ML-EID published somewhere? Thanks.

**Reply by abtink** [reply]
> When node is added to network isn't information about ML-EID published somewhere? Thanks

What you are asking is what DNS-SD (SRP) does :)... 

Device registers its service(s) and address(es) with an SRP server, then other device can query the server to discover the device and learn its  address(es).
